File: src/utils/geometry.py
# SYNAPSE - A Gateway of Intelligent Perception for Traffic Management
# Copyright (C) 2026 Noxfort Systems
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# File: src/utils/geometry.py
# Author: Gabriel Moraes
# Date: 2026-04-27
#
# Pure computational geometry functions.
# Extracted from FastMapMatcher to allow reuse across the codebase
# (e.g., CartographerAgent, spatial validators, etc.).

import math
import numpy as np
from typing import List, Tuple


def point_to_polyline_distance(px: float, py: float, polyline: List[Tuple[float, float]]) -> float:
    """
    Minimum perpendicular distance from point (px, py) to a polyline.
    Uses segment-by-segment projection with clamped parametric t.

    Args:
        px: X coordinate of the query point.
        py: Y coordinate of the query point.
        polyline: Ordered list of (x, y) vertices defining the polyline.

    Returns:
        Minimum distance in the same unit as the coordinates.
    """
    if not polyline:
        return float('inf')

    min_dist = float('inf')

    for i in range(len(polyline) - 1):
        ax, ay = polyline[i]
        bx, by = polyline[i + 1]

        # Vector AB
        abx = bx - ax
        aby = by - ay

        # Vector AP
        apx = px - ax
        apy = py - ay

        # Project AP onto AB, clamped to [0, 1]
        ab_sq = abx * abx + aby * aby
        if ab_sq < 1e-12:
            # Degenerate segment (zero length)
            dist = math.sqrt(apx * apx + apy * apy)
        else:
            t = max(0.0, min(1.0, (apx * abx + apy * aby) / ab_sq))
            # Closest point on segment
            cx = ax + t * abx
            cy = ay + t * aby
            dx = px - cx
            dy = py - cy
            dist = math.sqrt(dx * dx + dy * dy)

        min_dist = min(min_dist, dist)

    return min_dist


def frechet_distance(polyline_a: List[Tuple[float, float]], polyline_b: List[Tuple[float, float]]) -> float:
    """
    Discrete Fréchet distance between two polylines.
    Measures similarity between curves accounting for vertex ordering.

    Uses recursive DP for small inputs (n*m ≤ 10000) and switches to
    iterative DP for larger polylines to avoid stack overflow.

    Complexity: O(n·m) time and space.

    Args:
        polyline_a: First polyline as list of (x, y) tuples.
        polyline_b: Second polyline as list of (x, y) tuples.

    Returns:
        Fréchet distance in the same unit as the coordinates.
    """
    n = len(polyline_a)
    m = len(polyline_b)

    if n == 0 or m == 0:
        return float('inf')

    # Use iterative approach for large polylines to avoid stack overflow
    if n * m > 10000:
        return _frechet_iterative(polyline_a, polyline_b)

    # Dynamic programming table
    ca = np.full((n, m), -1.0)

    def _dist(i: int, j: int) -> float:
        dx = polyline_a[i][0] - polyline_b[j][0]
        dy = polyline_a[i][1] - polyline_b[j][1]
        return math.sqrt(dx * dx + dy * dy)

    def _frechet_rec(i: int, j: int) -> float:
        if ca[i, j] >= 0.0:
            return ca[i, j]

        d = _dist(i, j)

        if i == 0 and j == 0:
            ca[i, j] = d
        elif i == 0:
            ca[i, j] = max(_frechet_rec(0, j - 1), d)
        elif j == 0:
            ca[i, j] = max(_frechet_rec(i - 1, 0), d)
        else:
            ca[i, j] = max(
                min(
                    _frechet_rec(i - 1, j),
                    _frechet_rec(i - 1, j - 1),
                    _frechet_rec(i, j - 1)
                ),
                d
            )
        return ca[i, j]

    return _frechet_rec(n - 1, m - 1)


def _frechet_iterative(polyline_a: List[Tuple[float, float]], polyline_b: List[Tuple[float, float]]) -> float:
    """Iterative Fréchet distance for large polylines (avoids recursion limit)."""
    n = len(polyline_a)
    m = len(polyline_b)
    ca = np.zeros((n, m))

    def _dist(i, j):
        dx = polyline_a[i][0] - polyline_b[j][0]
        dy = polyline_a[i][1] - polyline_b[j][1]
        return math.sqrt(dx * dx + dy * dy)

    ca[0, 0] = _dist(0, 0)
    for i in range(1, n):
        ca[i, 0] = max(ca[i - 1, 0], _dist(i, 0))
    for j in range(1, m):
        ca[0, j] = max(ca[0, j - 1], _dist(0, j))

    for i in range(1, n):
        for j in range(1, m):
            ca[i, j] = max(
                min(ca[i - 1, j], ca[i - 1, j - 1], ca[i, j - 1]),
                _dist(i, j)
            )

    return ca[n - 1, m - 1]

File: ui/src_ui/utils/geometry.ts
// SYNAPSE - A Gateway of Intelligent Perception for Traffic Management
// Copyright (C) 2026 Noxfort Systems
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//
// File: ui/src_ui/utils/geometry.ts
// Author: Gabriel Moraes
// Date: 2026-08-29

import { Point2D, BoundingBox, MapNodeData } from '../types/topology';

/**
 * Calculates Euclidean distance from point p to line segment (v, w).
 */
export function distToSegment(p: Point2D, v: Point2D, w: Point2D): number {
  const l2 = (v.x - w.x) ** 2 + (v.y - w.y) ** 2;
  if (l2 === 0) return Math.hypot(p.x - v.x, p.y - v.y);
  let t = ((p.x - v.x) * (w.x - v.x) + (p.y - v.y) * (w.y - v.y)) / l2;
  t = Math.max(0, Math.min(1, t));
  return Math.hypot(p.x - (v.x + t * (w.x - v.x)), p.y - (v.y + t * (w.y - v.y)));
}

/**
 * Calculates the shortest distance from point p to any segment in a polyline.
 */
export function distToPolyline(p: Point2D, points: number[][]): number {
  if (!points || points.length === 0) return Infinity;
  if (points.length === 1) return Math.hypot(p.x - points[0][0], p.y - points[0][1]);

  let minDist = Infinity;
  for (let i = 0; i < points.length - 1; i++) {
    const v: Point2D = { x: points[i][0], y: points[i][1] };
    const w: Point2D = { x: points[i + 1][0], y: points[i + 1][1] };
    const d = distToSegment(p, v, w);
    if (d < minDist) {
      minDist = d;
    }
  }
  return minDist;
}


/**
 * Calculates the bounding box of a collection of 2D nodes.
 */
export function calculateBoundingBox(nodes: MapNodeData[]): BoundingBox {
  let minX = Infinity;
  let maxX = -Infinity;
  let minY = Infinity;
  let maxY = -Infinity;

  for (const n of nodes) {
    if (n.x < minX) minX = n.x;
    if (n.x > maxX) maxX = n.x;
    if (n.y < minY) minY = n.y;
    if (n.y > maxY) maxY = n.y;
  }

  return {
    minX: minX === Infinity ? 0 : minX,
    maxX: maxX === -Infinity ? 100 : maxX,
    minY: minY === Infinity ? 0 : minY,
    maxY: maxY === -Infinity ? 100 : maxY,
  };
}

/**
 * Calculates initial zoom and pan parameters to center and fit the map within viewport.
 */
export function calculateFitTransform(
  box: BoundingBox,
  viewportWidth: number,
  viewportHeight: number
): { zoom: number; pan: Point2D } {
  const rangeX = box.maxX - box.minX || 100;
  const rangeY = box.maxY - box.minY || 100;

  const scaleX = (viewportWidth * 0.8) / rangeX;
  const scaleY = (viewportHeight * 0.8) / rangeY;
  const initialZoom = Math.min(scaleX, scaleY, 5);

  const centerX = (box.minX + box.maxX) / 2;
  const centerY = (box.minY + box.maxY) / 2;

  return {
    zoom: initialZoom,
    pan: {
      x: viewportWidth / 2 - centerX * initialZoom,
      y: viewportHeight / 2 + centerY * initialZoom, // Invert SUMO Y axis
    },
  };
}

/**
 * Converts screen coordinates to SUMO Cartesian coordinates.
 */
export function screenToSumoCoords(screenPos: Point2D, pan: Point2D, zoom: number): Point2D {
  return {
    x: (screenPos.x - pan.x) / zoom,
    y: -(screenPos.y - pan.y) / zoom,
  };
}

/**
 * Samples a point at normalized fraction t (0 <= t <= 1) along a polyline.
 */
export function samplePolyline(points: number[][], t: number): [number, number] {
  if (!points || points.length === 0) return [0, 0];
  if (points.length === 1 || t <= 0) return [points[0][0], points[0][1]];
  if (t >= 1) return [points[points.length - 1][0], points[points.length - 1][1]];

  let totalLength = 0;
  const segLengths: number[] = [];
  for (let i = 0; i < points.length - 1; i++) {
    const len = Math.hypot(points[i + 1][0] - points[i][0], points[i + 1][1] - points[i][1]);
    segLengths.push(len);
    totalLength += len;
  }

  if (totalLength === 0) return [points[0][0], points[0][1]];

  const targetDist = t * totalLength;
  let accumulated = 0;

  for (let i = 0; i < segLengths.length; i++) {
    const nextAcc = accumulated + segLengths[i];
    if (targetDist <= nextAcc || i === segLengths.length - 1) {
      const segFraction = segLengths[i] > 0 ? (targetDist - accumulated) / segLengths[i] : 0;
      const x = points[i][0] + segFraction * (points[i + 1][0] - points[i][0]);
      const y = points[i][1] + segFraction * (points[i + 1][1] - points[i][1]);
      return [x, y];
    }
    accumulated = nextAcc;
  }

  return [points[points.length - 1][0], points[points.length - 1][1]];
}

/**
 * Computes the exact centerline polyline between two opposing lane polylines,
 * perfectly averaging bends and curves.
 */
export function computeCenterline(
  shapeA: number[][],
  shapeB: number[][],
  samples: number = 10
): number[][] {
  if (!shapeA || shapeA.length === 0) return shapeB || [];
  if (!shapeB || shapeB.length === 0) return shapeA;

  // shapeB runs in the opposite direction (from toNode to fromNode), so we reverse it
  const reversedB = [...shapeB].reverse();

  if (shapeA.length === 2 && reversedB.length === 2) {
    return [
      [(shapeA[0][0] + reversedB[0][0]) / 2, (shapeA[0][1] + reversedB[0][1]) / 2],
      [(shapeA[1][0] + reversedB[1][0]) / 2, (shapeA[1][1] + reversedB[1][1]) / 2],
    ];
  }

  const sampleCount = Math.max(samples, Math.max(shapeA.length, reversedB.length));
  const result: number[][] = [];

  for (let i = 0; i <= sampleCount; i++) {
    const t = i / sampleCount;
    const ptA = samplePolyline(shapeA, t);
    const ptB = samplePolyline(reversedB, t);
    result.push([(ptA[0] + ptB[0]) / 2, (ptA[1] + ptB[1]) / 2]);
  }

  return result;
}
